envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/EnvoyListenersFactory.kt

```diff
@@ -0,0 +1,298 @@
+
+        val defaultServiceTagFilterRules = serviceTagFilterRules()
+        val defaultHeaderToMetadataConfig = headerToMetadataConfig(defaultServiceTagFilterRules)
+        val defaultHeaderToMetadataFilter = { _: Group -> headerToMetadataHttpFilter(defaultHeaderToMetadataConfig) }
+        val defaultEnvoyRouterHttpFilter = { _: Group -> envoyRouterHttpFilter() }
+        val defaultApiConfigSource: ApiConfigSource = apiConfigSource()
+        val defaultEgressFilters = listOf(defaultHeaderToMetadataFilter, defaultEnvoyRouterHttpFilter)
+        val defaultIngressFilters = listOf(defaultEnvoyRouterHttpFilter)
+
+        private fun envoyRouterHttpFilter(): HttpFilter = HttpFilter.newBuilder().setName("envoy.router").build()
+
+        fun apiConfigSource(): ApiConfigSource {
+            return ApiConfigSource.newBuilder()
+                    .setApiType(ApiConfigSource.ApiType.GRPC)
+                    .addGrpcServices(GrpcService.newBuilder()
+                            .setEnvoyGrpc(
+                                    GrpcService.EnvoyGrpc.newBuilder()
+                                            .setClusterName("envoy-control-xds")
+                            )
+                    ).build()
+        }
+
+        fun headerToMetadataHttpFilter(headerToMetadataConfig: HeaderToMetadataConfig.Builder): HttpFilter {
+            return HttpFilter.newBuilder()
+                    .setName("envoy.filters.http.header_to_metadata")
+                    .setTypedConfig(ProtobufAny.pack(
+                            headerToMetadataConfig.build()
+                    ))
+                    .build()
+        }
+
+        fun headerToMetadataConfig(rules: List<HeaderToMetadataConfig.Rule>): HeaderToMetadataConfig.Builder {
+            val headerToMetadataConfig = HeaderToMetadataConfig.newBuilder()
+                    .addRequestRules(
+                            HeaderToMetadataConfig.Rule.newBuilder()
+                                    .setHeader("x-canary")
+                                    .setRemove(false)
+                                    .setOnHeaderPresent(
+                                            HeaderToMetadataConfig.KeyValuePair.newBuilder()
+                                                    .setKey("canary")
+                                                    .setMetadataNamespace("envoy.lb")
+                                                    .setType(HeaderToMetadataConfig.ValueType.STRING)
+                                                    .build()
+                                    )
+                                    .build()
+                    )
+
+            rules.forEach {
+                headerToMetadataConfig.addRequestRules(it)
+            }
+
+            return headerToMetadataConfig
+        }
+
+        fun serviceTagFilterRules(
+            header: String = "x-service-tag",
+            tag: String = "tag"
+        ): List<HeaderToMetadataConfig.Rule> {
```

Comment:
why list if it is always just one rule?

Reply:
Someone can use this class as to build their own serviceTagFilter that includes more than one Rule.